main.go

```diff
@@ -475,22 +475,93 @@ func makeSecretChanges(kubeClient *k8s.Client, secret *corev1.Secret, initiator
+}
 
-	status = "failed"
+func postEventAboutStatus(kubeClient *k8s.Client, secret *corev1.Secret, eventType string, action string, reason string, kind string, reportingController string, reportingInstance string)(err error){
+	log.Info().Msgf(" Starting Function")
+	now := time.Now()
+	secs := int64(now.Unix())
+	count := int32(1)
+	var eventResp corev1.Event
+	var exist bool
+	exist, err = isEventExist(kubeClient, *secret.Metadata.Namespace, *secret.Metadata.Name, &eventResp)
+	if exist {
+		count = *eventResp.Count+ 1
+		eventResp.Type = &eventType
+		eventResp.Action = &action
+		eventResp.Reason = &reason
+		eventResp.Count = &count
+		eventResp.LastTimestamp.Seconds = &secs
+		err = kubeClient.Update(context.Background(), &eventResp)
+		if err != nil {
+			log.Error().Msgf("Event %v.%v - Updating Event has an error.\n\t%s", *eventResp.Metadata.Name, *eventResp.Metadata.Namespace, err.Error())
+			return err
+		}
+		log.Info().Msgf("Event %v.%v - has been updated successfully...", *eventResp.Metadata.Name, *eventResp.Metadata.Namespace)
+		return
+	}
+	event := new(corev1.Event)
+	event.Metadata = new(metav1.ObjectMeta)
+	event.Metadata.Name = secret.Metadata.Name
+	event.Metadata.Namespace = secret.Metadata.Namespace
+	event.Metadata.CreationTimestamp = new(metav1.Time)
+	event.Metadata.CreationTimestamp.Seconds = &secs
+
+	event.Type = &eventType
+	event.Action = &action
+	event.Reason = &reason
+	event.Count = &count
+
+	event.FirstTimestamp = new(metav1.Time)
+	event.FirstTimestamp.Seconds = &secs
+
+	event.LastTimestamp = new(metav1.Time)
+	event.LastTimestamp.Seconds = &secs
 
+	event.InvolvedObject = new(corev1.ObjectReference)
+	event.InvolvedObject.Namespace = secret.Metadata.Namespace
+	event.InvolvedObject.Kind = &kind
+
+	event.EventTime = new(metav1.MicroTime)
+	event.EventTime.Seconds = &secs
+
+	event.ReportingComponent = &reportingController
+	event.ReportingInstance = &reportingInstance
+
+	err = kubeClient.Create(context.Background(), event)
+	if err != nil {
+		log.Error().Msgf("Event %v.%v - Creating Event has an error.\n\t%s", *eventResp.Metadata.Name, *eventResp.Metadata.Namespace, err.Error())
+		return err
+	}
+	log.Info().Msgf("Event %v.%v - has been created successfully...", *event.Metadata.Name, *event.Metadata.Namespace)
+	return
+}
+
+func processSecret(kubeClient *k8s.Client, secret *corev1.Secret, initiator string) (status string, err error) {
+	status = "failed"
 	if &secret != nil && &secret.Metadata != nil && &secret.Metadata.Annotations != nil {
 
 		desiredState := getDesiredSecretState(secret)
 		currentState := getCurrentSecretState(secret)
-
 		status, err = makeSecretChanges(kubeClient, secret, initiator, desiredState, currentState)
-
+		if status == "failed" {
+			err = postEventAboutStatus(kubeClient,secret, "Warning", strings.Title(status), "Letsencrypt Certificate has been created failed", "Secret","estafette.io/letsencrypt-certificate", "estafette-letsencrypt-certificate-9fdc7f864")
```

Comment:
I guess the last parameter should be dynamic?

Reply:
I think its better to be dynamic, the last parameter should I get the ID of the controller instance, e.g. `kubelet-xyzf`, What do you think the easiest way the get this value command line, or api?.